### PR TITLE
feat: add analytics client to alexa runtime (VF-1239)

### PR DESCRIPTION
**Pull request by xavidop**
**Fixes or implements VF-1239**

### Brief description. What is this change?

Added analytics system to this runtime

### Implementation details. How do you make this change?

The development changes a little bit from general runtime, it calls one per response and one per request.

### Checklist

- [x] title of PR reflects the branch name
- [x] all commits to adhere to conventional commits
- [x] appropriate tests have been written
- [ ] all the dependencies are upgraded

### CODE REVIEW

File: lib/clients/index.ts [hunk 1]
```diff
@@ -19,6 +20,7 @@ export interface ClientMap extends StaticType {
   metrics: MetricsType;
   mongo: MongoDB | null;
   pg: PostgresDB | null;
+  analyticsClient: AnalyticsSystem | null;
```
Comment (leartgjoni-voiceflow) [inline lib/clients/index.ts:23]:
```suggestion
  analyticsClient: AnalyticsSystem;
```
it seems like it's always being init (line 41 below), so no need for null

File: lib/services/alexa/request/lifecycle/initialize.ts [hunk 2]
```diff
@@ -23,6 +23,11 @@ const utilsObj = {
 export const initializeGenerator = (utils: typeof utilsObj) => async (runtime: AlexaRuntime, input: AlexaHandlerInput): Promise<void> => {
   const { requestEnvelope } = input;
 
+  // Identify on analytics system
+  if (runtime?.services?.AnalyticsClient) {
+    runtime.services.analyticsClient.identify(runtime.getVersionID());
+  }
+
```
Comment (leartgjoni-voiceflow) [inline lib/services/alexa/request/lifecycle/initialize.ts:30]:
```suggestion
    runtime.services.analyticsClient.identify(runtime.getVersionID());
```
as it's not null anymore

File: lib/services/alexa/request/lifecycle/response.ts [hunk 3]
```diff
@@ -51,6 +52,18 @@ export const responseGenerator = (utils: typeof utilsObj) => async (runtime: Ale
...
+    runtime.services.analyticsClient.track(
+      runtime.getVersionID(),
+      Event.INTERACT,
+      false,
+      response,
+      input.requestEnvelope.session?.sessionId,
+      runtime.getFinalState()
+    );
+  }
+
```
Comment (leartgjoni-voiceflow) [inline lib/services/alexa/request/lifecycle/response.ts:66]:
```suggestion
    runtime.services.analyticsClient.track(
      runtime.getVersionID(),
      Event.INTERACT,
      false,
      response,
      input.requestEnvelope.session?.sessionId,
      runtime.getFinalState()
    );
```

File: lib/services/alexa/request/lifecycle/update.ts [hunk 4]
```diff
@@ -15,11 +18,22 @@ const update = async (runtime: AlexaRuntime): Promise<void> => {
...
+  if (runtime?.services?.analyticsClient) {
+    runtime.services.analyticsClient.track(
+      runtime.getVersionID(),
+      Event.INTERACT,
+      true,
+      runtime.getRequest(),
+      input.requestEnvelope.session?.sessionId,
+      runtime.getFinalState()
+    );
+  }
```
Comment (leartgjoni-voiceflow) [inline lib/services/alexa/request/lifecycle/update.ts:36]: is it needed here too? It's already being called in `response.ts`. If not needed here, also revert the `input` param added to this func

**Review by leartgjoni-voiceflow — CHANGES_REQUESTED** [review]

File: lib/services/alexa/request/lifecycle/response.ts (hunk 3 above)
Comment (xavidop) [reply lib/services/alexa/request/lifecycle/response.ts]: done

File: lib/services/alexa/request/lifecycle/initialize.ts (hunk 2 above)
Comment (xavidop) [reply lib/services/alexa/request/lifecycle/initialize.ts]: done

File: lib/clients/index.ts (hunk 1 above)
Comment (xavidop) [reply lib/clients/index.ts]: done

File: tests/lib/services/alexa/request/lifecycle/initialize.unit.ts [hunk 5]
```diff
@@ -48,8 +50,13 @@ describe('initialize lifecycle unit tests', async () => {
         get: sinon.stub().returns(null),
       };
 
+      const config = {} as Config;
+
       const runtime = {
         getVersionID: sinon.stub().returns(VERSION_ID),
+        services: {
+          analyticsClient: AnalyticsClient(config),
+        },
```
Comment (leartgjoni-voiceflow) [inline tests/lib/services/alexa/request/lifecycle/initialize.unit.ts:59]: In the tests, each client or service injected needs to be stubbed, and then we assert that the stub was called correctly
```suggestion
      // declare stub
       const identifyStub = sinon.stub();
       // inject stub
        services: {
          analyticsClient: {identify: identifyStub},
        },
        // assert stub
        expect(identifyStub.args).to.eql([[VERSION_ID]]);
```

File: tests/lib/services/alexa/request/lifecycle/response.unit.ts [hunk 6]
```diff
@@ -18,13 +20,19 @@ describe('response lifecycle unit tests', () => {
...
       turn: { get: turnGet, set: sinon.stub() },
       stack: { isEmpty: sinon.stub().returns(true) },
       variables: { get: sinon.stub().returns(false) },
       getFinalState: sinon.stub().returns(finalState),
+      services: {
+        analyticsClient: AnalyticsClient(config),
+      },
+      getVersionID: () => {
+        return '';
+      },
```
Comment (leartgjoni-voiceflow) [inline tests/lib/services/alexa/request/lifecycle/response.unit.ts:35]: same as before this needs to be a stub. 
```suggestion
      getVersionID: sinon.stub().returns(VERSION_ID),
```

**Review by leartgjoni-voiceflow — CHANGES_REQUESTED** [review]

File: lib/services/alexa/request/lifecycle/update.ts (hunk 4 above)
Comment (xavidop) [reply lib/services/alexa/request/lifecycle/update.ts]: solved

File: tests/lib/services/alexa/request/lifecycle/response.unit.ts (hunk 6 above)
Comment (xavidop) [reply tests/lib/services/alexa/request/lifecycle/response.unit.ts]: done

File: tests/lib/services/alexa/request/lifecycle/initialize.unit.ts (hunk 5 above)
Comment (xavidop) [reply tests/lib/services/alexa/request/lifecycle/initialize.unit.ts]: done

**Review by leartgjoni-voiceflow — APPROVED** [review]

**Review by frankgu968 — APPROVED** [review]
lgtm!

File: lib/clients/ingest-client.ts [hunk 7]
```diff
@@ -24,6 +24,12 @@ export enum Event {
   INTERACT = 'interact',
 }
 
+export enum Request {
```
Comment (leartgjoni-voiceflow) [inline lib/clients/ingest-client.ts:27]: `RequestType` is a better name

File: lib/services/alexa/request/lifecycle/update.ts [hunk 8]
```diff
@@ -15,11 +17,24 @@ const update = async (runtime: AlexaRuntime): Promise<void> => {
...
+
+  let request: InteractRequestType = InteractRequestType.REQUEST;
+  if (input?.requestEnvelope?.request?.type === 'LaunchRequest') {
+    request = InteractRequestType.LAUNCH;
+  }
+  // Track response on analytics system
+  runtime.services.analyticsClient.track(
+    runtime.getVersionID(),
+    Event.INTERACT,
+    request,
```
Comment (leartgjoni-voiceflow) [inline lib/services/alexa/request/lifecycle/update.ts:33]:
```suggestion
    input?.requestEnvelope?.request?.type === 'LaunchRequest' ? InteractRequestType.LAUNCH : InteractRequestType.REQUEST,
```
we could do this inline as the `request` var is never being reused besides here

**Review by leartgjoni-voiceflow — CHANGES_REQUESTED** [review]

File: lib/services/alexa/request/lifecycle/update.ts (hunk 8 above)
Comment (xavidop) [reply lib/services/alexa/request/lifecycle/update.ts]: done

File: lib/clients/ingest-client.ts (hunk 7 above)
Comment (xavidop) [reply lib/clients/ingest-client.ts]: done

**Review by leartgjoni-voiceflow — APPROVED** [review]

**Review by jonahsnider — APPROVED** [review]

**Pull request merged**